aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/CrtResource.kt

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.sdk.kotlin.crt
+
+import kotlinx.cinterop.*
+import libcrt.*
+
+@OptIn(ExperimentalForeignApi::class)
+public abstract class CrtResource<T : CPointed> : CValuesRef<T>() {
+
+    public abstract val ptr: CPointer<T>
+
+    override fun getPointer(scope: AutofreeScope): CPointer<T> = ptr
+
+    private val rc = cValue<aws_ref_count>()
+
+    init {
+        aws_ref_count_init(
+            ref_count = rc,
+            `object` = ptr,
+            on_zero_fn = null,
```

Comment:
**correctness**: without a zero function we won't ever de-allocate anything. At least leave a TODO/FIXME here and we can revisit in memory management.